### PR TITLE
Support for Tensorflow model export

### PR DESCRIPTION
**Pull request by hardikdava**
Now, yolov8 models can be export to Tensorflow saved_model, keras, tflite (int8, int16, fp16, fp32) models.

Changes:

- updated some dependencies
- Export for following models possible:
  - Tensorflow saved_model
  - TFLite with int8, int16, fp16, fp32 formats
  - Keras model

Related issue: #198 

I have read the CLA Document and I sign the CLA.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR appears to be empty, containing no code changes.

### 📊 Key Changes
- There are no code changes to report in this PR.

### 🎯 Purpose & Impact
- Since the PR has no content, it has no direct purpose or impact on users as it stands.
- The PR might be in preparation for future changes, or it may have been opened by mistake. It could also be a placeholder for discussions or administrative purposes.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@hardikdava thanks for the PR! I think you might need to paste the CLA line without the period at the end. You just need to do it once and then it remembers for future PRs.

**Comment by hardikdava** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
@hardikdava testing this now.

**Comment by glenn-jocher** [comment]
@hardikdava I wasn't able to export to any TF format unfortunately. My code to reproduce the errors I'm seeing is below. I tested in our Colab notebook:

Screenshot 2023-01-10 at 21 23 23


```
# Git clone method (for development)
!git clone https://github.com/hardikdava/ultralytics -b patch-1
%pip install -qe ultralytics

!yolo mode=export model=yolov8n.pt format=saved_model
!yolo mode=export model=yolov8n.pt format=pb
!yolo mode=export model=yolov8n.pt format=tflite
```

Also the onnxsim optimization has been failing even before this PR, and I'm not sure why:
Screenshot 2023-01-10 at 21 22 33

**Comment by glenn-jocher** [comment]
@sergiossm tracking TF export issue

**Comment by hardikdava** [comment]
@glenn-jocher sorry for late reply. I checked with onnx2tf commands again. Here you can find my experimental [google colab](https://colab.research.google.com/drive/1cB-pagmj5KpUjhAQF3OTKFBqzAXG7Uvy?usp=sharing). You can find related references and comments in **export** section.

**Comment by torkiasalem** [comment]
@glenn-jocher  @AyushExel  I can not convert my yolov8 model from onnx to tenserflow or keras !! 

image

**Comment by glenn-jocher** [comment]
@torkiasalem, I'm sorry you're having trouble with model conversion. Can you please provide more details on the error messages you're seeing? In general, the conversion process from ONNX to TensorFlow or Keras should be straightforward. At a high level, you can use the `tf.keras.utils.convert_all_kernels_in_model` function to run the conversion on your ONNX model. Can you try to provide us with more information or the error messages you're encountering, so we can help further?

**Comment by h-pozuelo** [comment]
Can somebody help me to interpret the output Tensor???

I tried to do a print on the Tensor I got.

This is the console.log(obj.print())

Tensor
    [[[11.9871788 , 16.5831833, 20.122963 , ..., 590.333252 , 596.328125 , 604.4353027],
      [-0.0533504 , 27.0586033, 26.5866718, ..., 595.449707 , 602.3311768, 617.612793 ],
      [31.9734612 , 38.8257294, 47.6009521, ..., 117.8925171, 116.7957153, 90.239563  ],
      ...,
      [0          , 0         , 0         , ..., 0          , 0          , 3e-7       ],
      [0          , 0         , 0         , ..., 0          , 0          , 0          ],
      [0          , 0         , 0         , ..., 0          , 0          , 0.0000014  ]]]

**Comment by h-pozuelo** [comment]
@scraus 
Some people are telling me the following:

"" The output tensor from YOLOv8 ONNX model is basically everything included (bbox, score, and class). But you must do post process to get that because the output is really raw. YOLOv8 TF model isn't like YOLOv5 which is end2end model so the postprocess is needed. ""

The thing is that I dont know how to post proccess that output

**Comment by scraus** [comment]
@h-pozuelo is that also what you're figuring out right now?

**Comment by h-pozuelo** [comment]
@scraus 
I'm trying to. Don't know if I'm going able to achieve it

**Comment by scraus** [comment]
@h-pozuelo Are you considering using ONNX instead? I think this blog discusses the post-process: https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/230-yolov8-optimization/230-yolov8-optimization.ipynb

I have not tried it myself. If you have successfully achieved it please share! 🙏🏻

**Comment by h-pozuelo** [comment]
I think I'm gonna use onnx

**Comment by h-pozuelo** [comment]
I'm following this tutorial: https://dev.to/andreygermanov/how-to-detect-objects-in-videos-in-a-web-browser-using-yolov8-neural-network-and-javascript-lfb

But, because of my model being pre-trained with pyTorch+GPU, I don't know if its working (but very slow) or not working at all :-/

**Comment by h-pozuelo** [comment]
@glenn-jocher 

The issue I'm having with my tfjs model is that I dont know how to interpret the output tensor I get.
I don't know how to (from that single Tensor) obtain bbox, score, class, etc...

I tried a obj.print() of the Tensor, and the console log I got was this:

Tensor
    [[[11.9871788 , 16.5831833, 20.122963 , ..., 590.333252 , 596.328125 , 604.4353027],
      [-0.0533504 , 27.0586033, 26.5866718, ..., 595.449707 , 602.3311768, 617.612793 ],
      [31.9734612 , 38.8257294, 47.6009521, ..., 117.8925171, 116.7957153, 90.239563  ],
      ...,
      [0          , 0         , 0         , ..., 0          , 0          , 3e-7       ],
      [0          , 0         , 0         , ..., 0          , 0          , 0          ],
      [0          , 0         , 0         , ..., 0          , 0          , 0.0000014  ]]]
      
I don't even know if my tfjs model works , is fast or slow.... because Im not even able to obtain the data I need to draw the box

**Comment by glenn-jocher** [comment]
@scraus hello, thanks for reaching out. 

The output tensor from the YOLOv8 tfjs model is very "raw" and needs to be post-processed to obtain the bounding box, score and class values from it. The coordinates for the bounding box can be obtained by calculating the sigmoid of the tensor values, scaling them and applying an offset. Once you have the coordinates, you can overlay a rectangle on the image using the bounding box values. 

To optimize the performance of your tfjs YOLOv8 model, we recommend using ONNX format by following the instructions in the model conversion guide (https://github.com/ultralytics/ultralytics/blob/master/docs/tutorials/yolov8.md). This could improve the ability to use the model in-browser and gain better performance. 

If you're having trouble understanding the structure of the output tensor, you may be able to rely on the YOLOv8 community and Ultralytics team for guidance or check the repositories for documents/examples that demonstrate it. 

I hope this helps!
